crates/block-producer/src/block_producer.rs

```diff
@@ -220,6 +245,33 @@ impl BlockProducer {
         tx_skeleton
             .cell_deps_mut()
             .push(contracts_dep.omni_lock.clone().into());
+        // header_deps, used for obtaining l1 time in the state-validator-script
+        let block_number = block.raw().number().unpack();
+        if 2 <= rollup_context.determine_global_state_version(block_number) {
+            let finality_as_duration = rollup_context.rollup_config.finality_as_duration();
+            let last_finalized_timestamp = match Timepoint::from_full_value(
+                global_state.last_finalized_block_number().unpack(),
+            ) {
+                Timepoint::Timestamp(last_finalized_timestamp) => last_finalized_timestamp,
+                Timepoint::BlockNumber(_) => {
+                    unreachable!("2 <= global_state_version, last_finalized_timepoint should be timestamp-based");
+                }
+            };
+            let l1_confirmed_timestamp =
+                last_finalized_timestamp.saturating_add(finality_as_duration);
+            let l1_confirmed_header =
+                query_l1_header_by_timestamp(&self.rpc_client, l1_confirmed_timestamp)
+                    .await?
+                    .unwrap_or_else(|| {
+                        panic!(
+                            "get_l1_header_by_timestamp l1_timestamp={}, l2block={:#x}",
+                            l1_confirmed_timestamp, block_number
+                        )
+                    });
+            tx_skeleton
+                .header_deps_mut()
+                .push(ckb_types::prelude::Unpack::unpack(&l1_confirmed_header.hash()).pack());
```

Comment:
Header dep would need to be checked in psc or header dep error need to properly handled. We would need to revert the L2 block or at least re-generate the submission tx if the block is no longer on the main chain due to reorg.

Reply:
After you mentioned this question, I added [a code comment](https://github.com/godwokenrises/godwoken/pull/836/files#diff-224689380cb9536d4aec312f5f5092b209163247355a0f9e9435394606e92024R261-R264) to explain it:

```rust
            // Recover the l1_confirmed_header based on l1_confimed_timestamp.
            // The l1_confirmed_header is referenced when producing l2block and GlobalState.
            //
            // The block producer would reference confirmed headers to avoid L1 reorg issue.
```

Can this answer your question?

Reply:
Personally I think we still need to handle possible deep reorgs.

Reply:
For now we will handle this manually by reverting L2 blocks.

Reply:
Thanks for pointing this out. I have added your comment to block_producer.rs https://github.com/godwokenrises/godwoken/pull/836/files#diff-224689380cb9536d4aec312f5f5092b209163247355a0f9e9435394606e92024R183-R193:

```rust
        // By applying finality mechanism based on L1 timestamp, we assign the L1 timestamp,
        // obtained from a L1 header_dep, to GlobalState.last_finalized_timepoint.
        //
        // But L1 chain rollback makes things difficult. It must be a relation between
        // GlobalState or l2block and L1 header_dep, while L1 rollback, the relation should
        // be updated in time.
        //
        // Currently, we choose a L1 confirmed block header to be L1 header_dep, to avoid
        // L1 reorg issue.
        //
        // For deep reorg, we have to recover Godwoken by manually reverting L2.
```